lib/event_framework/event_store/sink.rb

```diff
@@ -68,29 +61,20 @@ def sink_staged_events(staged_events, database)
         new_event_rows
       end
 
-      def get_lock_with_retry(correlation_id:)
-        tries = 0
-        begin
-          lock_result = try_lock(database)
-          raise ConcurrencyError, "error obtaining lock" unless locked?(lock_result)
-        rescue ConcurrencyError => e
-          tries += 1
-          if tries > lock_retry_count
-            logger.info(msg: "event_framework.event_store.sink.max_retries_reached", tries: tries, correlation_id: correlation_id)
-            raise e
-          end
-          logger.info(msg: "event_framework.event_store.sink.retry", tries: tries, correlation_id: correlation_id)
-          sleep lock_retry_sleep
-          retry
+      # Set a local lock_timeout within a transaction then get an exclusive
+      # advisory lock so that we're the only database connection that can sink
+      # an event.
+      #
+      # If you're modifing the locking logic you can test that it's working
+      # correctly using the ./bin/demonstrate_event_sequence_id_gaps script.
+      def with_lock(correlation_id:)
+        database.transaction do
+          database.execute("SET LOCAL lock_timeout = '#{LOCK_TIMEOUT_MILLISECONDS}ms'; SELECT pg_advisory_xact_lock(-1)")
+          yield
         end
-      end
-
-      def try_lock(database)
-        database.select(Sequel.function(:pg_try_advisory_xact_lock, -1)).first
-      end
-
-      def locked?(lock_result)
-        lock_result && lock_result[:pg_try_advisory_xact_lock]
+      rescue Sequel::DatabaseLockTimeout => e
+        logger.info(msg: "event_framework.event_store.sink.lock_error", correlation_id: correlation_id)
+        raise ConcurrencyError, "error obtaining lock"
```

Comment:
I presume you don't want to change the exception class at this point since it may break users of the gem?

Reply:
Yeah exactly. Probably should be done fairly soon, but in a separate PR.